File: scatter_plot_app.py
"""
Shiny Scatter Plot Application in Python using plotnine.
"""

import pandas as pd
import numpy as np
from shiny import App, Inputs, Outputs, Session, module, reactive, render, ui
from shinywidgets import output_widget, render_widget
import plotly.express as px

from scatter import scatter_plotnine


def make_mock_df(n: int = 200, seed: int = 42) -> pd.DataFrame:
    """Generate mock dataset for generic scatter plot app."""
    np.random.seed(seed)
    df = pd.DataFrame({
        "subject": [f"ind_{i+1}" for i in range(n)],
        "x": np.random.randn(n),
        "sex": np.random.choice(["Female", "Male"], n),
        "diet": np.random.choice(["Chow", "HF"], n),
        "geno": np.random.choice(["AA", "AB", "BB"], n),
    })
    y = (
        2 * df["x"]
        + np.where(df["sex"] == "Male", 1.5, 0.0)
        + np.where(df["diet"] == "HF", -1.0, 0.0)
        + np.where(df["geno"] == "AB", 0.5, np.where(df["geno"] == "BB", 1.0, 0.0))
        + np.random.randn(n) * 0.8
    )
    df["y"] = y
    df.index = df["subject"]
    return df


@module.ui
def scatter_plot_input():
    return ui.output_ui("scatter_inputs")


@module.ui
def scatter_plot_output():
    return ui.output_ui("plot_ui")


@module.server
def scatter_plot_server(
    input: Inputs,
    output: Outputs,
    session: Session,
    plot_df: reactive.Calc,
    x_label: callable = lambda: "x",
    y_label: callable = lambda: "y"
):
    @output
    @render.ui
    def scatter_inputs():
        dat = plot_df()
        if dat is None or dat.empty:
            return ui.TagList()

        cols = list(dat.columns)
        candidate_vars = [c for c in cols if c not in ["x", "y", "subject"]]

        if "sex" in cols and "diet" in cols:
            candidate_vars.append("sex_diet")

        color_choices = {"none": "None"}
        for v in candidate_vars:
            color_choices[v] = v

        shape_choices = {"none": "None"}
        for v in candidate_vars:
            if v != "sex_diet":
                shape_choices[v] = v

        facet_choices = {"none": "None"}
        for v in candidate_vars:
            facet_choices[v] = v

        return ui.TagList(
            ui.input_radio_buttons("static", "Plot Type:", ["Static", "Interactive"], selected="Static"),
            ui.input_select("color_by", "Color by:", choices=color_choices, selected="none"),
            ui.input_select("shape_by", "Shape by:", choices=shape_choices, selected="none"),
            ui.input_select("facet_by", "Facet by:", choices=facet_choices, selected="none"),
            ui.input_checkbox("add_line", "Add regression line?", value=True),
            ui.input_slider("point_size", "Point Size:", min=1, max=10, value=3, step=0.5),
            ui.input_slider("point_alpha", "Transparency (Alpha):", min=0.1, max=1, value=0.7, step=0.1),
        )

    @reactive.calc
    def plot_obj():
        dat = plot_df()
        if dat is None or "x" not in dat.columns or "y" not in dat.columns:
            return None

        col_by = input.color_by() if "color_by" in input else "none"
        shp_by = input.shape_by() if "shape_by" in input else "none"
        fct_by = input.facet_by() if "facet_by" in input else "none"
        add_l = input.add_line() if "add_line" in input else True
        pt_sz = input.point_size() if "point_size" in input else 3
        pt_al = input.point_alpha() if "point_alpha" in input else 0.7

        x_lbl = x_label() if callable(x_label) else x_label
        y_lbl = y_label() if callable(y_label) else y_label

        return scatter_plotnine(
            dat=dat,
            color_by=col_by,
            shape_by=shp_by,
            facet_by=fct_by,
            add_line=add_l,
            point_size=pt_sz,
            point_alpha=pt_al,
            x_label=x_lbl,
            y_label=y_lbl
        )

    @output
    @render.plot
    def plot_static():
        p = plot_obj()
        return p

    @output
    @render_widget
    def plot_interactive():
        dat = plot_df()
        if dat is None or dat.empty:
            return None
        df = dat.copy()
        if "sex" in df.columns and "diet" in df.columns:
            df["sex_diet"] = df["sex"].astype(str) + "_" + df["diet"].astype(str)

        col_by = input.color_by() if ("color_by" in input and input.color_by() != "none") else None
        shp_by = input.shape_by() if ("shape_by" in input and input.shape_by() != "none") else None
        fct_by = input.facet_by() if ("facet_by" in input and input.facet_by() != "none") else None

        x_lbl = x_label() if callable(x_label) else x_label
        y_lbl = y_label() if callable(y_label) else y_label

        add_l = input.add_line() if "add_line" in input else True

        open_symbols = ["circle-open", "square-open", "diamond-open", "triangle-up-open", "triangle-down-open", "star-open", "cross-open"]
        fig = px.scatter(
            df,
            x="x",
            y="y",
            color=col_by,
            symbol=shp_by,
            symbol_sequence=open_symbols,
            facet_col=fct_by,
            hover_name="subject" if "subject" in df.columns else None,
            labels={"x": x_lbl, "y": y_lbl},
            trendline="ols" if add_l else None
        )
        if not shp_by:
            fig.update_traces(marker_symbol="circle-open")
        fig.update_layout(template="plotly_white")
        return fig

    @output
    @render.ui
    def plot_ui():
        static_val = input.static() if "static" in input else "Static"
        if static_val == "Interactive":
            return output_widget("plot_interactive")
        else:
            return ui.output_plot("plot_static")


def scatter_plot_app():
    """Create and return the Shiny App for scatter plot."""
    mock_df = make_mock_df()

    app_ui = ui.page_sidebar(
        ui.sidebar(
            scatter_plot_input("scatter")
        ),
        ui.card(
            scatter_plot_output("scatter")
        ),
        title="Test Generic Scatter Plot (Python / plotnine)"
    )

    def server(input: Inputs, output: Outputs, session: Session):
        scatter_plot_server("scatter", reactive.calc(lambda: mock_df))

    return App(app_ui, server)


app = scatter_plot_app()

File: scatterPlotApp.R
#' Shiny Scatter Plot App
#'
#' @param id identifier for shiny reactive
#' @param plot_df reactive data frame containing columns: x, y, and optionally sex, diet, geno
#'
#' @author Brian S Yandell, \email{brian.yandell@@wisc.edu}
#' @keywords utilities
#'
#' @return No return value; called for side effects.
#'
#' @export
#' @importFrom ggplot2 ggplot aes geom_point geom_smooth theme_minimal scale_color_brewer scale_color_hue facet_wrap labs vars scale_shape_manual
#' @importFrom plotly renderPlotly plotlyOutput ggplotly
#' @importFrom shiny checkboxInput fluidRow column isolate isTruthy moduleServer NS observeEvent plotOutput radioButtons reactive renderPlot renderUI req selectInput sliderInput strong tagList uiOutput updateSelectInput updateSliderInput withProgress
#' @importFrom bslib card layout_sidebar page_sidebar sidebar
source("scatter.R")
source("theme.R")

scatterPlotApp <- function() {
  # Mock data frame creation
  set.seed(42)
  n <- 200
  mock_df <- data.frame(
    subject = paste0("ind_", 1:n),
    x = rnorm(n),
    sex = sample(c("Female", "Male"), n, replace = TRUE),
    diet = sample(c("Chow", "HF"), n, replace = TRUE),
    geno = sample(c("AA", "AB", "BB"), n, replace = TRUE),
    stringsAsFactors = FALSE
  )
  # y is a function of x, sex, diet, and genotype
  mock_df$y <- 2 * mock_df$x +
    ifelse(mock_df$sex == "Male", 1.5, 0) +
    ifelse(mock_df$diet == "HF", -1, 0) +
    ifelse(mock_df$geno == "AB", 0.5, ifelse(mock_df$geno == "BB", 1, 0)) +
    rnorm(n, sd = 0.8)

  rownames(mock_df) <- mock_df$subject

  ui <- bslib::page_sidebar(
    title = "Test Generic Scatter Plot",
    sidebar = bslib::sidebar(
      scatterPlotInput("scatter")
    ),
    bslib::card(scatterPlotOutput("scatter"))
  )

  server <- function(input, output, session) {
    scatterPlotServer("scatter", shiny::reactive(mock_df))
  }

  shiny::shinyApp(ui, server)
}
#' @export
#' @rdname scatterPlotApp
scatterPlotServer <- function(id, plot_df, x_label = shiny::reactive("x"), y_label = shiny::reactive("y")) {
  shiny::moduleServer(id, function(input, output, session) {
    ns <- session$ns

    # Render dynamic UI choices based on columns of plot_df
    output$scatter_inputs <- shiny::renderUI({
      dat <- shiny::req(plot_df())
      cols <- colnames(dat)

      # Exclude x, y, and subject for color/shape/facet variables
      candidate_vars <- cols[!(cols %in% c("x", "y", "subject"))]

      # If sex and diet are present, sex_diet is a candidate
      if (all(c("sex", "diet") %in% cols)) {
        candidate_vars <- c(candidate_vars, "sex_diet")
      }

      # Clean choices
      color_choices <- c("None" = "none")
      for (var in candidate_vars) {
        color_choices[var] <- var
      }

      shape_choices <- c("None" = "none")
      for (var in candidate_vars[candidate_vars != "sex_diet"]) {
        shape_choices[var] <- var
      }

      facet_choices <- c("None" = "none")
      for (var in candidate_vars) {
        facet_choices[var] <- var
      }

      shiny::tagList(
        shiny::radioButtons(ns("static"), "Plot Type:", c("Static", "Interactive"), "Static"),
        shiny::selectInput(ns("color_by"), "Color by:", choices = color_choices, selected = "none"),
        shiny::selectInput(ns("shape_by"), "Shape by:", choices = shape_choices, selected = "none"),
        shiny::selectInput(ns("facet_by"), "Facet by:", choices = facet_choices, selected = "none"),
        shiny::checkboxInput(ns("add_line"), "Add regression line?", TRUE),
        shiny::sliderInput(ns("point_size"), "Point Size:", min = 1, max = 10, value = 3, step = 0.5),
        shiny::sliderInput(ns("point_alpha"), "Transparency (Alpha):", min = 0.1, max = 1, value = 0.7, step = 0.1)
      )
    })

    # Render static ggplot
    output$plot_static <- shiny::renderPlot({
      print(shiny::req(plot_obj()))
    })

    # Render interactive plotly
    output$plot_interactive <- plotly::renderPlotly({
      p <- shiny::req(plot_obj())
      plotly::ggplotly(p)
    })

    # Renders the UI wrapper that switches static/interactive
    output$plot_ui <- shiny::renderUI({
      static_val <- ifelse(is.null(input$static), "Static", input$static)
      switch(static_val,
        Static = shiny::plotOutput(ns("plot_static")),
        Interactive = plotly::plotlyOutput(ns("plot_interactive"))
      )
    })

    # Construct the plot object
    plot_obj <- shiny::reactive({
      scatter_ggplot(
        dat = shiny::req(plot_df()),
        color_by = input$color_by,
        shape_by = input$shape_by,
        facet_by = input$facet_by,
        add_line = input$add_line,
        point_size = input$point_size,
        point_alpha = input$point_alpha,
        x_label = x_label(),
        y_label = y_label()
      )
    })

    # Return the ggplot reactive
    plot_obj
  })
}
#' @export
#' @rdname scatterPlotApp
scatterPlotInput <- function(id) {
  ns <- shiny::NS(id)
  shiny::uiOutput(ns("scatter_inputs"))
}
#' @export
#' @rdname scatterPlotApp
scatterPlotOutput <- function(id) {
  ns <- shiny::NS(id)
  shiny::uiOutput(ns("plot_ui"))
}
